### PR TITLE
user pref to filter pre-release helm charts

### PR DESCRIPTION
**Pull request by mantis-toboggan-md**
#2280 
Screen Shot 2021-03-10 at 7 12 40 AM
Screen Shot 2021-03-10 at 7 12 02 AM

### CODE REVIEW

**Review by lvuch — APPROVED** [review]
lgtm

File: pages/c/_cluster/apps/charts.vue [hunk 1]
```diff
@@ -313,6 +327,10 @@ export default {
         btnCb(false);
       }
     },
+
+    isPreRelease(version = '') {
+      return semver.prerelease(version) || version.includes('-rc');
```
Comment (vincent99) [inline pages/c/_cluster/apps/charts.vue:332]: This isn't the right definition of prerelease.  See https://semver.org/ # 9 (but just contains hyphen doesn't work either, see 10).  The semver parser probably does this?

File: pages/c/_cluster/apps/charts.vue [hunk 2]
```diff
@@ -344,8 +362,9 @@ export default {
           :class="{'pull-left': true, 'repo': true, [r.color]: true}"
           @input="toggleRepo(r, $event)"
         />
-      </div>
 
+        <Checkbox v-model="showPrerelease" class="repo mt-3" :label="t('catalog.charts.showPreRelease')" />
```
Comment (vincent99) [inline pages/c/_cluster/apps/charts.vue:366]: I don't really get the point of the checkbox here, unless a chart has ONLY prereleases it won't change the list of whole charts displayed.

**Review by vincent99 — CHANGES_REQUESTED** [review]
Checking if an upgrade is available on the installed apps screen needs to be handled too.

As Adrian noted prerelease is off and pretty well hidden by default in Helm (the flag is called--devel) so I'm wondering if it's worth even showing the checkbox in 2-3 separate screens vs just putting it in the regular user prefs page once as "Show Helm Charts: [Releases only | Include Prerelease Versions ]" instead.

File: pages/c/_cluster/apps/charts.vue (hunk 2 above)
Comment (mantis-toboggan-md) [reply pages/c/_cluster/apps/charts.vue]: That's fair. If this isn't relevant often I think it makes sense to move the toggle to the prefs page.

File: pages/c/_cluster/apps/charts.vue (hunk 1 above)
Comment (mantis-toboggan-md) [reply pages/c/_cluster/apps/charts.vue]: This parser seems to have a pretty narrow definition of 'prerelease' which includes '-alpha' and '-beta' but not '-rc'; updated based off this definition:  

> ...a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty.

File: pages/c/_cluster/apps/install.vue [hunk 3]
```diff
@@ -492,8 +494,15 @@ export default {
             nue.disabled = true;
           }
         }
+        if (!this.showPrerelease) {
+          const isPre = !!(version.version || '').match(/.[0-9]+(-[0-9A-Za-z-]+)/);
```
Comment (vincent99) [inline pages/c/_cluster/apps/install.vue:498]: This and the other one in charts.vue should be combined so the logic for "is <str> a prerelease" is in only one place (utils/version.js would be a good one).

Metadata (after the `+`) can contain numbers and hyphens too so this doesn't quite work (e.g. `v1.18.6+k3s1-1` will match but is not a prerelease).  You'd have to split off the `+` section first then see if the remainder has a `-`.

But.. you said the library only looks for alpha/beta/etc, which library is that?  The `semver` we have already loaded seems to work?  It's a whole formal generated grammar even, and the words alpha/beta don't appear in it, outside of comments:

```javascript
> const x = require('semver');
> semver.parse('v1.2.3-foo1+k3s2-2');
SemVer {
  options: { loose: false, includePrerelease: false },
  loose: false,
  raw: 'v1.2.3-foo1+k3s2-2',
  major: 1,
  minor: 2,
  patch: 3,
  prerelease: [ 'foo1' ],
  build: [ 'k3s2-2' ],
  version: '1.2.3-foo1'
}
> semver.prerelase('v1.2.3-foo1+bar2-3');
[ 'foo1' ]
> semver.prerelease('v1.2.3');
null
```

**Comment by vincent99** [comment]
Merged via #2688

**Pull request closed without merging**